sys/xtimer/xtimer.c

```diff
@@ -58,46 +58,51 @@ void _xtimer_sleep(uint32_t offset, uint32_t long_offset)
     mutex_lock(&mutex);
 }
 
-void xtimer_usleep_until(uint32_t *last_wakeup, uint32_t interval) {
+void xtimer_periodic_wakeup(uint32_t *last_wakeup, uint32_t period) {
     xtimer_t timer;
     mutex_t mutex = MUTEX_INIT;
 
     timer.callback = _callback_unlock_mutex;
     timer.arg = (void*) &mutex;
 
-    uint32_t target = *last_wakeup + interval;
-
+    uint32_t target = (*last_wakeup) + period;
     uint32_t now = xtimer_now();
     /* make sure we're not setting a value in the past */
-    if (now < *last_wakeup) {
-        /* base timer overflowed */
-        if (!((target < *last_wakeup) && (target > now))) {
+    if (now < (*last_wakeup)) {
+        /* base timer overflowed between last_wakeup and now */
+        if (!((now < target) && (target < (*last_wakeup)))) {
+            /* target time has already passed */
             goto out;
         }
     }
-    else if (! ((target < *last_wakeup) || (target > now))) {
-        goto out;
+    else {
+        /* base timer did not overflow */
+        if ((((*last_wakeup) < target) && (target < now))) {
+            /* target time has already passed */
+            goto out;
+        }
     }
 
-    /* For large offsets, set an absolute target time, as
-     * it is more exact.
+    /*
+     * For large offsets, set an absolute target time.
+     * As that might cause an underflow, for small offsets, spin.
      *
-     * As that might cause an underflow, for small offsets,
-     * use a relative offset, as that can never underflow.
+     * Note: last_wakeup _must never_ specify a time in the future after
+     * _xtimer_periodic_sleep returns.
+     * If this happens, last_wakeup may specify a time in the future when the
+     * next call to _xtimer_periodic_sleep is made, which in turn will trigger
+     * the overflow logic above and make the next timer fire too early, causing
+     * last_wakeup to point even further into the future, leading to a chain
+     * reaction.
      *
-     * For very small offsets, spin.
+     * tl;dr Don't return too early!
      */
     uint32_t offset = target - now;
-
-    if (offset > (XTIMER_BACKOFF * 2)) {
```

Comment:
I'm a little reluctant to just drop the extra safety, because I did do quite some experimenting to get this stable (non-underflowing). But I'm ok with the other changes.

Would you mind splitting this out of this PR?